File: 7-reverse-integer/reverse-integer.py
class Solution:
    def reverse(self, x: int) -> int:
        revNum = 0
        intMax = 2**31 - 1
        intMin = -2**31

        isNegative = x < 0
        x = abs(x)

        while x > 0:
            lastDigit = x % 10
            x = x // 10
            revNum = revNum * 10 + lastDigit

        if isNegative:
            revNum = -revNum

        # Check overflow after reversing
        if revNum < intMin or revNum > intMax:
            return 0

        return revNum

File: 7-reverse-integer/reverse-integer.java
class Solution {
    public int reverse(int x) {
        long rev = 0; // use long to catch overflow
        int sign = x < 0 ? -1 : 1;
        x = Math.abs(x);

        while (x > 0) {
            rev = rev * 10 + x % 10;
            x /= 10;
        }
        rev *= sign;

        return (rev < Integer.MIN_VALUE || rev > Integer.MAX_VALUE) ? 0 : (int) rev;
    }
}
